### ISSUE TITLE
Input/output error on `elfloader` when running `chmod` syscall on a directory

### ISSUE THREAD

**Issue by John-Ted** [unikraft/unikraft#812]
### Describe the bug

When running a program in `elfloader` that does a `chmod` syscall on a directory, the syscall returns `errno 5`, Input/output error.

### Steps to reproduce

Run the chmod program in https://github.com/unikraft/static-pie-apps/pull/84 with the provided rootfs mounted.
Config file for `elfloader`: [example_config.txt](https://github.com/unikraft/unikraft/files/11069837/example_config.txt)

### Expected behavior

I expected the syscall to succeed.

### Relevant log output

```shell
[...]
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
             Epimetheus 0.12.0~5bd4b94d
[    0.628200] dbg:  [libukboot] <boot.c @  365> Call constructor: 0x19b3a0()...
[    0.629301] dbg:  [appelfloader] <main.c @   86> Searching for image...
[    0.630319] dbg:  [libuksched] <thread.c @  254> uk_thread 0x40013a018 (elfapp): ctx:0x40013a018, ectx:0x40013b1a0, tlsp:0x40013b190
[    0.632159] dbg:  [libcontext] <tls.c @  173> tls_area_init: target: 0x40013b020 (376 bytes)
[    0.633459] dbg:  [libcontext] <tls.c @  175> tls_area_init: copy (.tdata): 4 bytes
[    0.634636] dbg:  [libcontext] <tls.c @  177> tls_area_init: uninitialized (.tbss): 364 bytes
[    0.635952] dbg:  [libcontext] <tls.c @  179> tls_area_init: pad: 0 bytes
[    0.636995] dbg:  [libcontext] <tls.c @  181> tls_area_init: tcb: 8 bytes
[    0.638037] dbg:  [libcontext] <tls.c @  183> tls_area_init: tcb self ptr: 0x40013b190
[    0.639259] dbg:  [libcontext] <tls.c @  204> (tls_area): 40013b020  0d f0 b0 b0 00 00 00 00  00 00 00 00 00 00 00 00  |................|
[    0.641183] dbg:  [libcontext] <tls.c @  204> (tls_area): 40013b030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
[    0.643103] dbg:  [libcontext] <tls.c @  204> *
[    0.643796] dbg:  [libcontext] <tls.c @  204> (tls_area): 40013b190  90 b1 13 00 04 00 00 00                           |........|
[    0.645595] dbg:  [libuksched] <thread.c @  144> uk_thread 0x40013a018 (elfapp) init: Call initialization 0x126ed0()...
[    0.647262] dbg:  [libposix_process] <process.c @  414> thread 0x40013a018 (elfapp): Parent 0x5fec29018 (init) has no PID, skipping...
[    0.649132] dbg:  [appelfloader] <main.c @  115> Load image...
[    0.650035] dbg:  [appelfloader] <elf_load.c @  108> 0x219000: ELF machine type: 62
[    0.651219] dbg:  [appelfloader] <elf_load.c @  124> 0x219000: ELF OS ABI: 3
[    0.652300] dbg:  [appelfloader] <elf_load.c @  136> 0x219000: ELF object type: 3
[    0.653456] dbg:  [appelfloader] <elf_load.c @  186> 0x219000: phdr[0]: R--, offset: 0, vaddr: 0, paddr: 0, filesz: 32768 B, memsz 32768 B, align: 4096 B
[    0.655575] dbg:  [appelfloader] <elf_load.c @  197> 0x219000: \_ segment at pie + 0x0 (len: 0x8000) from file @  0x0 (len: 0x8000)
[    0.657397] dbg:  [appelfloader] <elf_load.c @  214> 0x219000: \_ base: pie + 0x0 (len: 0x8000)
[    0.658739] dbg:  [appelfloader] <elf_load.c @  186> 0x219000: phdr[1]: R-X, offset: 0x8000, vaddr: 0x8000, paddr: 0x8000, filesz: 619965 B, memsz 619965 B, align: 4096 B
[    0.661099] dbg:  [appelfloader] <elf_load.c @  197> 0x219000: \_ segment at pie + 0x8000 (len: 0x975bd) from file @  0x8000 (len: 0x975bd)
[    0.663038] dbg:  [appelfloader] <elf_load.c @  214> 0x219000: \_ base: pie + 0x0 (len: 0x9f5bd)
[    0.664386] dbg:  [appelfloader] <elf_load.c @  186> 0x219000: phdr[2]: R--, offset: 0xa0000, vaddr: 0xa0000, paddr: 0xa0000, filesz: 174198 B, memsz 174198 B, align: 4096 B
[    0.666772] dbg:  [appelfloader] <elf_load.c @  197> 0x219000: \_ segment at pie + 0xa0000 (len: 0x2a876) from file @  0xa0000 (len: 0x2a876)
[    0.668723] dbg:  [appelfloader] <elf_load.c @  214> 0x219000: \_ base: pie + 0x0 (len: 0xca876)
[    0.670074] dbg:  [appelfloader] <elf_load.c @  186> 0x219000: phdr[3]: RW-, offset: 0xcb510, vaddr: 0xcc510, paddr: 0xcc510, filesz: 23712 B, memsz 46576 B, align: 4096 B
[    0.672432] dbg:  [appelfloader] <elf_load.c @  197> 0x219000: \_ segment at pie + 0xcc510 (len: 0xb5f0) from file @  0xcb510 (len: 0x5ca0)
[    0.674365] dbg:  [appelfloader] <elf_load.c @  214> 0x219000: \_ base: pie + 0x0 (len: 0xd7b00)
[    0.675720] dbg:  [appelfloader] <elf_load.c @  242> 0x219000: Program/Library memory region: 0x400201000-0x4002d8b00
[    0.677349] dbg:  [appelfloader] <elf_load.c @  264> 0x219000: Copying 0x219000 - 0x221000 -> 0x400201000 - 0x400209000
[    0.679031] dbg:  [appelfloader] <elf_load.c @  274> 0x219000: Zeroing 0x400209000 - 0x400209000
[    0.680383] dbg:  [appelfloader] <elf_load.c @  288> 0x219000: Protecting 0x400201000 - 0x400209000: R--
[    0.681840] dbg:  [appelfloader] <elf_load.c @  264> 0x219000: Copying 0x221000 - 0x2b85bd -> 0x400209000 - 0x4002a05bd
[    0.683878] dbg:  [appelfloader] <elf_load.c @  274> 0x219000: Zeroing 0x4002a05bd - 0x4002a05bd
[    0.685228] dbg:  [appelfloader] <elf_load.c @  288> 0x219000: Protecting 0x400209000 - 0x4002a1000: R-X
[    0.686689] dbg:  [appelfloader] <elf_load.c @  264> 0x219000: Copying 0x2b9000 - 0x2e3876 -> 0x4002a1000 - 0x4002cb876
[    0.688460] dbg:  [appelfloader] <elf_load.c @  274> 0x219000: Zeroing 0x4002cb876 - 0x4002cb876
[    0.689811] dbg:  [appelfloader] <elf_load.c @  288> 0x219000: Protecting 0x4002a1000 - 0x4002cc000: R--
[    0.691275] dbg:  [appelfloader] <elf_load.c @  264> 0x219000: Copying 0x2e4510 - 0x2ea1b0 -> 0x4002cd510 - 0x4002d31b0
[    0.692945] dbg:  [appelfloader] <elf_load.c @  274> 0x219000: Zeroing 0x4002d31b0 - 0x4002d8b00
[    0.694306] dbg:  [appelfloader] <elf_load.c @  288> 0x219000: Protecting 0x4002cd000 - 0x4002d9000: RW-
[    0.695761] dbg:  [appelfloader] <main.c @  131> Prepare application thread...
[    0.696870] dbg:  [appelfloader] <elf_ctx.c @  107> net_all_info: image:          0x400201000 - 0x4002d8b00
[    0.698366] dbg:  [appelfloader] <elf_ctx.c @  109> net_all_info: start:          0x400201000
[    0.699671] dbg:  [appelfloader] <elf_ctx.c @  111> net_all_info: entry:          0x4002097a0
[    0.700980] dbg:  [appelfloader] <elf_ctx.c @  113> net_all_info: ehdr_phoff:     0x40
[    0.702197] dbg:  [appelfloader] <elf_ctx.c @  115> net_all_info: ehdr_phnum:     12
[    0.703386] dbg:  [appelfloader] <elf_ctx.c @  117> net_all_info: ehdr_phentsize: 0x38
[    0.704600] dbg:  [libcontext] <ctx.c @   69> ukarch_ctx 0x40013a018: start:0x4002097a0 sp:0x40014fe50
[    0.706034] dbg:  [libposix_process] <process.c @  173> Process PID 0: New thread TID 0
[    0.707264] dbg:  [libposix_process] <process.c @  281> Process PID 0 created (parent PID: -1)
[    0.708587] dbg:  [libposix_time] <time.c @   65> (int) uk_syscall_r_nanosleep((const struct timespec*) 0x311ee0, (struct timespec*) 0x311ee0)
[    0.710581] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "arch_prctl" (158) at ip:0x40020b10b (arg0=0x3001, arg1=0x40014fde0, ...)
[    0.712859] dbg:  [appelfloader] <arch_prctl.c @  113> (long) uk_syscall_r_arch_prctl((long) 0x3001, (long) 0x40014fde0, (long) 0x40025e640)
[    0.714796] dbg:  [appelfloader] <arch_prctl.c @  169> arch_prctl option code 0x3001 ignored
arch_prctl(0x3001, 0x40014fde0, ...) = Invalid argument (-22)
[    0.717423] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "brk" (12) at ip:0x400257ceb (arg0=0x0, arg1=0x40014fc88, ...)
[    0.719549] dbg:  [appelfloader] <brk.c @   56> (void *) uk_syscall_r_brk((void *) 0x0)
[    0.720777] dbg:  [appelfloader] <brk.c @   89> brk @ 0x5fe800000 (brk heap region: 0x5fe800000-0x5fea00000)
brk(NULL) = va:0x5fe800000
[    0.723127] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "brk" (12) at ip:0x400257ceb (arg0=0x5fe800dc0, arg1=0x40014fc88, ...)
[    0.725365] dbg:  [appelfloader] <brk.c @   56> (void *) uk_syscall_r_brk((void *) 0x5fe800dc0)
[    0.726699] dbg:  [appelfloader] <brk.c @   83> zeroing 0x5fe800000-0x5fe800dc0...
[    0.727865] dbg:  [appelfloader] <brk.c @   89> brk @ 0x5fe800dc0 (brk heap region: 0x5fe800000-0x5fea00000)
brk(va:0x5fe800dc0) = va:0x5fe800dc0
[    0.730453] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "arch_prctl" (158) at ip:0x40020c292 (arg0=0x1002, arg1=0x5fe8003c0, ...)
[    0.732732] dbg:  [appelfloader] <arch_prctl.c @  113> (long) uk_syscall_r_arch_prctl((long) 0x1002, (long) 0x5fe8003c0, (long) 0x4002d28e0)
[    0.734671] dbg:  [appelfloader] <arch_prctl.c @  123> arch_prctl option SET_FS(0x5fe8003c0)
arch_prctl(0x1002, 0x5fe8003c0, ...) = 0
[    0.737003] dbg:  [libsyscall_shim] <uk_syscall_binary.c @  122> System call updated userland TLS pointer register to 0x40013b190 (before: 0x5fe8003c0)
[    0.739092] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "set_tid_address" (218) at ip:0x40025ad7e (arg0=0x5fe800690, arg1=0x5fe8003c0, ...)
[    0.741501] dbg:  [libposix_futex] <futex.c @  328> (pid_t) uk_syscall_r_set_tid_address((pid_t *) 0x5fe800690)
[    0.743053] dbg:  [libposix_process] <process.c @  523> (pid_t) uk_syscall_r_gettid()
set_tid_address(0x5fe800690, 0x5fe8003c0, ...) = 0
[    0.745427] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "set_robust_list" (273) at ip:0x40025adda (arg0=0x5fe8006a0, arg1=0x18, ...)
[    0.747743] dbg:  [libsyscall_shim] <uk_syscall6_r.c @  944> syscall "set_robust_list" is not available
set_robust_list(0x5fe8006a0, 0x18, ...) = Function not implemented (-38)
[    0.750657] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "rseq" (334) at ip:0x40025ae11 (arg0=0x5fe800d60, arg1=0x20, ...)
[    0.752827] dbg:  [libsyscall_shim] <uk_syscall6_r.c @  944> syscall "rseq" is not available
rseq(0x5fe800d60, 0x20, ...) = Function not implemented (-38)
[    0.755447] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "uname" (63) at ip:0x40028264b (arg0=0x40014fb70, arg1=0x20, ...)
[    0.757618] dbg:  [libposix_sysinfo] <sysinfo.c @  140> (int) uk_syscall_r_uname((struct utsname *) 0x40014fb70)
uname(0x40014fb70, 0x20, ...) = 0
[    0.760137] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "prlimit64" (302) at ip:0x400257cb4 (arg0=0x0, arg1=0x3, ...)
[    0.762250] dbg:  [libposix_process] <deprecated.c @  317> (int) uk_syscall_r_prlimit64((int) 0x0, (unsigned int) 0x3, (struct rlimit *) 0x0, (struct rlimit *) 0x40014fcf0)
[    0.764624] dbg:  [libposix_process] <deprecated.c @  378> Resource 3: cur = 65536, max = 65536
prlimit64(0x0, 0x3, ...) = 0
[    0.766843] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "readlink" (89) at ip:0x400289a63 (arg0=0x4002b7ed8, arg1=0x40014ec30, ...)
[    0.769158] dbg:  [libvfscore] <main.c @ 2287> (ssize_t) uk_syscall_e_readlink((const char *) 0x4002b7ed8, (char *) 0x40014ec30, (size_t) 0x1000)
[    0.771163] dbg:  [libuk9p] <9p.c @  222> TWALK fid 0 newfid 2 nwname 1 name proc
[    0.772313] dbg:  [libkvmvirtio] <virtio_ring.c @  367> Old head:0, new head:2, total_desc:2
[    0.773622] dbg:  [libkvmvirtio9p] <virtio_9p.c @  285> notify queue 0
[    0.774679] dbg:  [libuk9p] <9preq.c @  226> RERROR No such file or directory 2
readlink("/proc/self/exe", "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"..., 4096) = No such file or directory (-2)
[    0.778692] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "getrandom" (318) at ip:0x40026e1c7 (arg0=0x4002d80b0, arg1=0x8, ...)
[    0.780920] dbg:  [libukswrand] <getrandom.c @   39> (ssize_t) uk_syscall_r_getrandom((void *) 0x4002d80b0, (size_t) 0x8, (unsigned int) 0x1)
getrandom(0x4002d80b0, 0x8, ...) = 8
[    0.783860] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "brk" (12) at ip:0x400257ceb (arg0=0x5fe821dc0, arg1=0x4002d1700, ...)
[    0.786100] dbg:  [appelfloader] <brk.c @   56> (void *) uk_syscall_r_brk((void *) 0x5fe821dc0)
[    0.787439] dbg:  [appelfloader] <brk.c @   83> zeroing 0x5fe800dc0-0x5fe821dc0...
[    0.788671] dbg:  [appelfloader] <brk.c @   89> brk @ 0x5fe821dc0 (brk heap region: 0x5fe800000-0x5fea00000)
brk(va:0x5fe821dc0) = va:0x5fe821dc0
[    0.791269] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "brk" (12) at ip:0x400257ceb (arg0=0x5fe822000, arg1=0x0, ...)
[    0.793399] dbg:  [appelfloader] <brk.c @   56> (void *) uk_syscall_r_brk((void *) 0x5fe822000)
[    0.794735] dbg:  [appelfloader] <brk.c @   83> zeroing 0x5fe821dc0-0x5fe822000...
[    0.795900] dbg:  [appelfloader] <brk.c @   89> brk @ 0x5fe822000 (brk heap region: 0x5fe800000-0x5fea00000)
brk(va:0x5fe822000) = va:0x5fe822000
[    0.798492] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "mprotect" (10) at ip:0x400257edb (arg0=0x4002cd000, arg1=0x4000, ...)
[    0.800737] dbg:  [libposix_mmap] <mmap.c @  227> (int) uk_syscall_r_mprotect((void *) 0x4002cd000, (size_t) 0x4000, (int) 0x1)
mprotect(va:0x4002cd000, 16384, PROT_READ) = OK
[    0.803856] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "chmod" (90) at ip:0x40025703b (arg0=0x4002a1004, arg1=0x1c0, ...)
[    0.806043] dbg:  [libvfscore] <main.c @ 2521> (int) uk_syscall_r_chmod((const char*) 0x4002a1004, (mode_t) 0x1c0)
[    0.807640] dbg:  [libuk9p] <9p.c @  222> TWALK fid 0 newfid 2 nwname 1 name chmod_test_dir
[    0.808924] dbg:  [libkvmvirtio] <virtio_ring.c @  367> Old head:0, new head:2, total_desc:2
[    0.810223] dbg:  [libkvmvirtio9p] <virtio_9p.c @  285> notify queue 0
[    0.811276] dbg:  [libuk9p] <9p.c @  244> RWALK nwqid 1
[    0.812077] dbg:  [libuk9p] <9p.c @  516> TSTAT fid 2
[    0.812849] dbg:  [libkvmvirtio] <virtio_ring.c @  367> Old head:0, new head:2, total_desc:2
[    0.814149] dbg:  [libkvmvirtio9p] <virtio_9p.c @  285> notify queue 0
[    0.815191] dbg:  [libuk9p] <9p.c @  524> RSTAT
[    0.815886] dbg:  [libuk9p] <9p.c @  516> TSTAT fid 2
[    0.816658] dbg:  [libkvmvirtio] <virtio_ring.c @  367> Old head:0, new head:2, total_desc:2
[    0.817959] dbg:  [libkvmvirtio9p] <virtio_9p.c @  285> notify queue 0
[    0.818997] dbg:  [libuk9p] <9p.c @  524> RSTAT
[    0.819689] dbg:  [libuk9p] <9p.c @  544> TWSTAT fid 2
[    0.820474] dbg:  [libkvmvirtio] <virtio_ring.c @  367> Old head:0, new head:2, total_desc:2
[    0.821774] dbg:  [libkvmvirtio9p] <virtio_9p.c @  285> notify queue 0
[    0.822811] dbg:  [libuk9p] <9preq.c @  226> RERROR Input/output error 5
[    0.823842] dbg:  [libuk9p] <9p.c @  427> TCLUNK fid 2
[    0.824632] dbg:  [libkvmvirtio] <virtio_ring.c @  367> Old head:0, new head:2, total_desc:2
[    0.825936] dbg:  [libkvmvirtio9p] <virtio_9p.c @  285> notify queue 0
[    0.826963] dbg:  [libuk9p] <9p.c @  431> RCLUNK
chmod(0x4002a1004, 0x1c0, ...) = Input/output error (-5)
[    0.828928] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "newfstatat" (262) at ip:0x400256ffe (arg0=0x1, arg1=0x4002b86dd, ...)
[    0.831170] dbg:  [libvfscore] <main.c @  913> (int) uk_syscall_r_newfstatat((int) 0x1, (const char*) 0x4002b86dd, (struct stat*) 0x40014f4c0, (int) 0x1000)
[    0.833334] dbg:  [libvfscore] <main.c @  871> (int) uk_syscall_e_fstat((int) 0x1, (struct stat *) 0x40014f4c0)
newfstatat(0x1, 0x4002b86dd, ...) = 0
[    0.835892] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "write" (1) at ip:0x400257247 (arg0=0x1, arg1=0x5fe801270, ...)
[    0.838035] dbg:  [libvfscore] <main.c @  732> (ssize_t) uk_syscall_r_write((int) 0x1, (const void *) 0x5fe801270, (size_t) 0x15)
[    0.839831] dbg:  [libvfscore] <main.c @  686> (ssize_t) uk_syscall_r_writev((int) 0x1, (const struct iovec *) 0x40014f520, (int) 0x1)
chmod failed errno=5
write(fd:1, "chmod failed errno=5\x0A", 21) = 21
[    0.843362] dbg:  [libsyscall_shim] <uk_syscall_binary.c @   88> Binary system call request "exit_group" (231) at ip:0x4002567b1 (arg0=0x0, arg1=0xe7, ...)
[    0.845506] dbg:  [libposix_process] <process.c @  564> (int) uk_syscall_r_exit_group((int) 0x0)
[    0.846857] dbg:  [libposix_process] <process.c @  367> Terminating PID 0: Self-killing TID 0...
[    0.848210] dbg:  [libuksched] <sched.c @  284> 0x400003018: thread 0x40013a018 (elfapp) terminated
[    0.849607] dbg:  [libuksched] <thread.c @  205> uk_thread 0x40013a018 (elfapp) term: Call termination 0x128990()...
[    0.851234] dbg:  [libuksched] <thread.c @  205> uk_thread 0x40013a018 (elfapp) term: Call termination 0x12d670()...
[    0.852853] dbg:  [libuksched] <thread.c @  205> uk_thread 0x40013a018 (elfapp) term: Call termination 0x1271c0()...
[    0.854474] dbg:  [libposix_process] <process.c @  448> thread 0x40013a018 (elfapp): Releasing thread with TID: 0 (PID: 0)
[    0.856179] dbg:  [libposix_process] <process.c @  322> Process PID 0 released
[    0.857288] dbg:  [libuksched] <sched.c @  294> 0x400003018: thread 0x40013a018 (elfapp) on gc list
[    0.858685] dbg:  [libuksched] <sched.c @  255> 0x400003018: garbage collect thread 0x40013a018 (elfapp)
Console terminated, terminating guest (PID: 261557)...
```

**Comment by Rohan-Jagannathan** [comment]
Hello, I am representing a group of undergraduate UT Austin students aiming to contribute to open-source virtualization projects on behalf of a virtualization course. Could we have this issue assigned to us?

**Comment by marcrittinghaus** [comment]
Hey @John-Ted, @andraprs 

I looked into this issue and know where the problem is. There are actually two issues here:

+ our driver deletes the DIRECTORY bit when setting the mode. This causes the `-EIO`
+ if you force the correct bits via `gdb` for the mode, the call still fails with `EISDIR`. I stepped through QEMU's virtio-9p device. The problem is that the `stat` struct that we send to change the mode is first retrieved, then modified to apply the changes and then sent back. However, this will provide a value for the `length` field in the `stat` struct. This in turn causes QEMU to try to `truncate` the directory fd, which does not work. Thus `EISDIR`. 

Actually, looking at the code, you are supposed to set fields that should not be touched to `-1` instead of loading their current value. Otherwise, QEMU will perform syscalls for all fields (e.g., to change the access time, rename, ...) even if they have not changed.

I will fix this and provide a PR for this.
